compilers/javascript/src/main/java/org/thingml/compilers/javascript/JSCompiler.java

```diff
@@ -1,153 +1,153 @@
+import org.thingml.compilers.thing.common.NewFSMBasedThingImplCompiler;
+import org.thingml.compilers.utils.OpaqueThingMLCompiler;
+import org.thingml.utilities.logging.Logger;
+import org.thingml.xtext.constraints.ThingMLHelpers;
+import org.thingml.xtext.helpers.AnnotatedElementHelper;
+import org.thingml.xtext.helpers.ConfigurationHelper;
+import org.thingml.xtext.thingML.Configuration;
+import org.thingml.xtext.thingML.Enumeration;
+import org.thingml.xtext.thingML.EnumerationLiteral;
+import org.thingml.xtext.thingML.Message;
+import org.thingml.xtext.thingML.Parameter;
+import org.thingml.xtext.thingML.Thing;
+import org.thingml.xtext.thingML.ThingMLModel;
+import org.thingml.xtext.thingML.Type;
+
+public abstract class JSCompiler extends OpaqueThingMLCompiler {
+
+	public JSCompiler(ThingActionCompiler thingActionCompiler, ThingApiCompiler thingApiCompiler,
+			CfgMainGenerator mainCompiler, CfgBuildCompiler cfgBuildCompiler, NewFSMBasedThingImplCompiler thingImplCompiler) {
+		super(thingActionCompiler, thingApiCompiler, mainCompiler, cfgBuildCompiler, thingImplCompiler);
+		this.ctx = new JSContext(this);
+	}
+
+	@Override
+    public boolean do_call_compiler(Configuration cfg, Logger log, String... options) {
+        ctx.addContextAnnotation("thisRef", "this.");
+        if (AnnotatedElementHelper.hasFlag(cfg, "use_fifo")) {
+        	ctx.addContextAnnotation("use_fifo", "true");	
+        }
+        //new File(ctx.getOutputDirectory() + "/" + cfg.getName()).mkdirs();
+        ctx.setCurrentConfiguration(cfg);
+        compile(cfg, ThingMLHelpers.findContainingModel(cfg), true, ctx);
+        ctx.getCompiler().getCfgBuildCompiler().generateDockerFile(cfg, ctx);
+        ctx.getCompiler().getCfgBuildCompiler().generateBuildScript(cfg, ctx);
+        ctx.writeGeneratedCodeToFiles();
+        ctx.generateNetworkLibs(cfg);
+        
+        return true;
+    }
+	
+	protected void exports(StringBuilder events, String e) {
+		events.append("exports." + e + " = " + e + ";\n\n");
+	}
+	
+	protected void generateEvents(Configuration cfg) {
+		StringBuilder events = ctx.getBuilder("events.js");								
+		for(Message m : ConfigurationHelper.allMessages(cfg)) {
+			String className = ctx.firstToUpper(m.getName())+'_'+ctx.firstToUpper(ThingMLHelpers.findContainingThing(m).getName());
+			String type = m.getName();
+			events.append("var " + className + " = /** @class */ (function () {\n");
+			events.append("  function " + className + "(from, port,...params) {\n");
+			events.append("    this.type = '" + type + "';\n");
+			events.append("    this.port = port;\n");
+			events.append("    this.from = from;\n");
+			StringBuilder params = new StringBuilder();
+			for(Parameter p : m.getParameters()) {
+				events.append("    this." + p.getName() + " = params[" + m.getParameters().indexOf(p) + "];\n");
+				if (m.getParameters().indexOf(p) > 0)
+					params.append(" + ', ' + ");
+				params.append('this.' + p.getName());

```

Comment:
@fungiboletus that should fix the bug you found in the toString (see line 95 below)